sim/battle-stream.ts

```diff
@@ -250,9 +255,9 @@ export class BattlePlayer {
 		this.stream = playerStream;
 		this.log = [];
 		this.debug = debug;
-		this.listen();
```

Comment:
If we want to maintain the 'start on creation' approach instead of an explicit `start` call (which I've renamed from `listen`, but I'd be happy to change back to `listen`), I would instead suggest we make the constructor private and add a static factory method which can then return a started instance and the promise in case someone cares:

```javascript
const {p1, started} = BattlePlayer.create(...);
// started.catch(err => ...)  if they want
```

Reply:
I wonder if we just make `BattlePlayer` extend a `Stream` and just foward errors to itself?

Reply:
Hmm. What would it stream other than errors? And how would using it look? `new BattlePlayer().catch(err => ...)` instead of `new BattlePlayer.start().catch(err => ...)`? I'm not sure it buys a whole lot, and is somewhat confusing logically (a player seems like something which would act on the stream, not be a stream itself?) but could work.

Reply:
It could also be a native Node EventEmitter with error events.

Reply:
I'd actually rather stay away from that route if possible. I'd like to be able to run the entirety of `sim` purely in browser for my '[PocketMon](https://docs.google.com/document/d/1xKy1b1SExXqlb6KXfmfbuifVCjAgPUbnusekIZuKQ8s/edit#heading=h.jb61j01iy1m4)' implementation)

Reply:
That's currently supported? I'm pretty surprised, but that makes me happy.

Reply:
Making `BattlePlayer` extend `Promise` might also be an option, but that might be excessive.

Reply:
I had it working at one point in my prototype, yeah (with some shimming etc). I mean, I bet a lot more could work given the low amount of dependencies (I could probably just change some of the socket stuff slightly and get it to play well as well?), but I did have two random bots playing each other on an android web view on my phone.